pkg/kubelet/kubelet.go

```diff
@@ -929,34 +929,107 @@ type Kubelet struct {
 	runtimeCache    kubecontainer.RuntimeCache
 	kubeClient      clientset.Interface
 	heartbeatClient clientset.Interface
-	rootDirectory   string
+	// mirrorPodClient is used to create and delete mirror pods in the API for static
+	// pods.
+	mirrorPodClient kubepod.MirrorClient
+
+	rootDirectory string
 
 	lastObservedNodeAddressesMux sync.RWMutex
 	lastObservedNodeAddresses    []v1.NodeAddress
 
 	// onRepeatedHeartbeatFailure is called when a heartbeat operation fails more than once. optional.
 	onRepeatedHeartbeatFailure func()
 
-	// podWorkers handle syncing Pods in response to events.
+	// podManager stores the desired set of admitted pods and mirror pods that the kubelet should be
+	// running. The actual set of running pods is stored on the podWorkers. The manager is populated
+	// by the kubelet config loops and may be consulted by other components that need to see the set
+	// of desired pods. Note that not all desired pods are running, and not all running pods are in
+	// the podManager - for instance, force deleting a pod from the apiserver will remove it from
+	// the podManager, but the pod may still be terminating and tracked by the podWorkers. Components
+	// that need to know the actual consumed resources of the node or are driven by podWorkers and
+	// the sync*Pod methods	(status, volume, stats) should also consult the podWorkers when
+	// reconciling.
+	//
+	// TODO: review all kubelet components that need the actual set of pods (vs the desired set)
+	// and update them to use podWorkers instead of podManager. This may introduce latency in some
+	// methods, but avoids race conditions and correctly accounts for terminating pods that have
+	// been force deleted or static pods that have been updated.
+	podManager kubepod.Manager
+
+	// podWorkers is responsible for driving the lifecycle state machine of each pod. The worker is
+	// notified of config changes, updates, periodic reconciliation, container runtime updates, and
+	// evictions of all desired pods and will invoke reconciliation methods per pod in separate
+	// goroutines. The podWorkers are authoritative in the kubelet for what pods are actually being
+	// run and their current state:
+	//
+	// * syncing: pod should be running (syncPod)
+	// * terminating: pod should be stopped (syncTerminatingPod)
+	// * terminated: pod should have all resources cleaned up (syncTerminatedPod)
```

Comment:
nit: It might be worth to note here that for the three states outlined here it is expected that invariant mentioned is only valid**after** the function is completed without error (not before or during).
 
For example, only if syncTermiantedPod completes without error, then we can assume that all `pod should have all resources cleaned up`.